File: csharp/855.exam-room.cs
/*
 * @lc app=leetcode id=855 lang=csharp
 *
 * [855] Exam Room
 */

using System.Collections.Generic;

// @lc code=start
public class ExamRoom {

    private int Count;
    private SortedSet<int> Sorted;

    public ExamRoom(int N) {
        Sorted = new SortedSet<int>();
        Count = N;
    }
    
    public int Seat() {
        int max = 0, pos = 0, prev = -1;
        foreach (int student in Sorted) {
            if (prev < 0) {
                max = student;
            } else {
                int dist = (student - prev) / 2;
                if (dist > max) {
                    max = dist;
                    pos = prev + dist;
                }
            }
            prev = student;
        }
        if (prev >= 0 && Count - prev - 1 > max) {
            pos = Count - 1;
        }
        Sorted.Add(pos);
        return pos;
    }
    
    public void Leave(int p) {
        Sorted.Remove(p);
    }
}

/**
 * Your ExamRoom object will be instantiated and called as such:
 * ExamRoom obj = new ExamRoom(N);
 * int param_1 = obj.Seat();
 * obj.Leave(p);
 */
// @lc code=end

File: java/855.exam-room.java
/*
 * @lc app=leetcode id=855 lang=java
 *
 * [855] Exam Room
 */

// @lc code=start
class ExamRoom {

    private int n;
    private Set<Integer> sorted;

    public ExamRoom(int N) {
        n = N;
        sorted = new TreeSet<Integer>();
    }
    
    public int seat() {
        int ret = 0, max = 0, prev = -1;
        for (int s : sorted) {
            if (prev == -1) {
                max = s;
            } else {
                int d = (s - prev) / 2;
                if (d > max) {
                    max = d;
                    ret = prev + d;
                }
            }
            prev = s;
        }
        if (prev != -1 && n - 1 - prev > max) {
            ret = n - 1;
        }
        sorted.add(ret);
        return ret;
    }
    
    public void leave(int p) {
        sorted.remove(p);
    }
}

/**
 * Your ExamRoom object will be instantiated and called as such:
 * ExamRoom obj = new ExamRoom(N);
 * int param_1 = obj.seat();
 * obj.leave(p);
 */
// @lc code=end

